src/main/java/de/muenchen/isi/domain/service/AbfrageStatusService.java

```diff
@@ -18,42 +30,209 @@
+     * Erstellt die Statemachine und initialisiert sie mit dem Status der Abfrage aus der DB.
+     * <p>
+     * Ausserdem definieren wir eine preStateChange Listener der bei einer Statusänderung den Status in der DB aktualiesiert.
      *
-     * @param abfrage vom Typ {@link AbfrageModel} zur Freigabe.
-     * @return das freigegebene {@link AbfrageModel}.
-     * @throws AbfrageStatusNotAllowedException falls die Abfrage nicht freigegeben werden kann.
+     * @param id vom Typ {@link UUID}  um die Abfrage aus der DB zu hollen.
+     * @throws EntityNotFoundException falls die Abfrage nicht gefunden werden kann.
      */
-    protected AbfrageModel freigabeAbfrage(final AbfrageModel abfrage, final UUID id) throws AbfrageStatusNotAllowedException {
-        if (Objects.equals(abfrage.getStatusAbfrage(), StatusAbfrage.ANGELEGT)) {
-            abfrage.setStatusAbfrage(StatusAbfrage.OFFEN);
-            return abfrage;
-        } else {
-            final var message = String.format(
-                    "Abfrage mit ID %s kann nicht freigegeben werden, da diese sich im Status %s befindet.",
-                    id.toString(),
-                    abfrage.getStatusAbfrage()
-            );
-            log.error(message);
-            throw new AbfrageStatusNotAllowedException(message);
+    private StateMachine<StatusAbfrage, StatusAbfrageEvents> build(final UUID id) throws EntityNotFoundException {
+        final InfrastrukturabfrageModel abfrage = abfrageService.getInfrastrukturabfrageById(id);
+        // Hier habe ich ein EntityNotFoundExeption Array definiert damit ich die Fehlermeldung im preStateChange Listener abfangen kann und
+        // danach werfen kann. Da man in den Lambdas keine Variablen zuweisen kann.
+        final EntityNotFoundException[] entityNotFoundException = new EntityNotFoundException[1];
+        StateMachine<StatusAbfrage, StatusAbfrageEvents> stateMachine = stateMachineFactory.getStateMachine(abfrage.getId());
+
+        stateMachine.stopReactively().block();
+
+        stateMachine.getStateMachineAccessor()
+                .doWithAllRegions(sma -> {
+
+                    sma.addStateMachineInterceptor(new StateMachineInterceptorAdapter<StatusAbfrage, StatusAbfrageEvents>() {
+
+                        @Override
+                        public void preStateChange(State<StatusAbfrage, StatusAbfrageEvents> state, Message<StatusAbfrageEvents> message,
+                                                   Transition<StatusAbfrage, StatusAbfrageEvents> transition, StateMachine<StatusAbfrage, StatusAbfrageEvents> stateMachine,
+                                                   StateMachine<StatusAbfrage, StatusAbfrageEvents> rootStateMachine) {
+                            Optional.ofNullable(message).ifPresent(msg -> {
+                                Optional.ofNullable(
+                                                UUID.class.cast(msg.getHeaders().getOrDefault(ABFRAGE_ID_HEADER, UUID.randomUUID())))
```

Comment:
Es wird in diesem Code das Objekt des Headers "abfrage_id" herausgeholt und dann gecastet. Falls der Header nicht exitiert wird eine zufällige UUID erzeugt.

Les- und Wartbarer wäre doch evtl folgendes:

```java

                                        final UUID abfrageId;
                                        final MessageHeaders headers = msg.getHeaders();
                                        if (headers.containsKey(ABFRAGE_ID_HEADER)) {
                                            abfrageId = (UUID) headers.get(ABFRAGE_ID_HEADER);
                                        } else {
                                            throw EineException();
                                        }

```

Reply:
Evtl in eigene Methode auslagern um den Code in dieser Klasse besser lesen und auch testen zu können.

Für den ausgelagertn Code Unit-Tests definieren.